aten/src/ATen/native/mps/operations/Bucketization.mm

```diff
@@ -0,0 +1,377 @@
+
+      // this function call is a no-op if MPS Profiler is not enabled
+      getMPSProfiler().beginProfileKernel(bucketizationPSO, kernel, {input, boundaries, sorter});
+
+      [computeEncoder setComputePipelineState:bucketizationPSO];
+      [computeEncoder setBuffer:inputBuffer offset:input.storage_offset() * input.element_size() atIndex:0];
+      [computeEncoder setBuffer:boundariesBuffer
+                         offset:boundaries.storage_offset() * boundaries.element_size()
+                        atIndex:1];
+      [computeEncoder setBuffer:outputBuffer offset:result.storage_offset() * result.element_size() atIndex:2];
+      [computeEncoder setBytes:&idim_in length:sizeof(int64_t) atIndex:3];
+      [computeEncoder setBytes:&idim_bd length:sizeof(int64_t) atIndex:4];
+      [computeEncoder setBytes:&numel_in length:sizeof(int64_t) atIndex:5];
+      [computeEncoder setBytes:&right_i64 length:sizeof(int64_t) atIndex:6];
+      [computeEncoder setBytes:&is_1d_boundaries length:sizeof(int64_t) atIndex:7];
+      if (sorter.defined())
+        [computeEncoder setBuffer:sorterBuffer offset:sorter.storage_offset() * sorter.element_size() atIndex:8];
+
+      // A threadGroup is equivalent to a cuda's block.
+      int64_t maxThreadgroups = 1024;
+      int64_t maxThreads = bucketizationPSO.maxTotalThreadsPerThreadgroup;
+      NSUInteger tgSize = std::min(maxThreads, numel_in);
+      MTLSize threadgroupsPerGrid = MTLSizeMake(std::min(maxThreadgroups, ceil_div<int64_t>(numel_in, tgSize)), 1, 1);
+      MTLSize threadGroupSize = MTLSizeMake(tgSize, 1, 1);
+      [computeEncoder dispatchThreadgroups:threadgroupsPerGrid threadsPerThreadgroup:threadGroupSize];
+
+      getMPSProfiler().endProfileKernel(bucketizationPSO);
+    }
+  });
+}
+} // namespace mps
+
+Tensor& searchsorted_out_mps(const Tensor& sorted_sequence,
+                             const Tensor& self,
+                             bool out_int32,
+                             bool right,
+                             const c10::optional<c10::string_view> side_opt,
+                             const c10::optional<Tensor>& sorter_opt,
+                             Tensor& result) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> sorter_maybe_owned = at::borrow_from_optional_tensor(sorter_opt);
+  const Tensor& sorter = *sorter_maybe_owned;
+  searchsorted_pre_check(sorted_sequence, self, result, out_int32, right, side_opt, sorter);
+  resize_output(result, self.sizes());
+
+  // we have two inputs to set right, pre_check checks that they aren't set to opposites
+  bool is_right = (side_opt && *side_opt == "right") || right;
+  if (self.numel() == 0) {
+    return result;
+  }
+
+  // for non-contiguous result tensors, we write the output to a contiguous copy so we can later copy back, maintaing
+  // the original result tensor
+  Tensor out = result;
+  if (!result.is_contiguous()) {
+    out = result.contiguous();
+  }
```

Comment:
Nit (also, I wonder if `t.contiguous()` just returns reference to self if tensor is already contagious, in which case one does not even need a ternary)
```suggestion
  Tensor out = result.is_contiguous() ? result : result.contiguous();
```

Reply:
Yes it does.

https://github.com/pytorch/pytorch/blob/24b61a45c9e8eeb7b7bee23ad8173e41f1e59d54/aten/src/ATen/core/TensorBase.h#L124C2-L130

But we do this contiguous check, which is unnecessary, in a lot of places in the codebase. I think it would be nice to correct them all together in a separate PR?